File: apps/mobile/android/app/src/main/kotlin/dev/anvil/PlatformEvent.kt
package dev.anvil

import org.json.JSONArray
import org.json.JSONObject

/**
 * Events pushed to the Rust core, normalised so that a Kotlin
 * `WifiAwareSession` callback and a Swift `NWBrowser` result arrive looking
 * identical (§9, §90).
 *
 * JSON rather than a generated binding, for the same reason as the Flutter
 * boundary: the volume is low, and a boundary a human can read in a log is
 * worth more here than the microseconds a binary encoding would save. Media
 * never crosses this boundary as JSON — audio frames go over a direct buffer
 * handoff.
 */
sealed class PlatformEvent {

    abstract fun toJson(): String

    data class PeerAdvertised(
        val kind: String,
        val handle: String,
        val address: String,
        val payload: ByteArray,
    ) : PlatformEvent() {
        override fun toJson(): String = JSONObject().apply {
            put("type", "peerAdvertised")
            put("kind", kind)
            put("handle", handle)
            put("address", address)
            put("payload", JSONArray(payload.map { it.toInt() and 0xFF }))
        }.toString()
    }

    data class PeerAdvertisementLost(val kind: String, val handle: String) : PlatformEvent() {
        override fun toJson(): String = JSONObject().apply {
            put("type", "peerAdvertisementLost")
            put("kind", kind)
            put("handle", handle)
        }.toString()
    }

    data class PathEstablished(val pathId: Long, val maxDatagramSize: Int) : PlatformEvent() {
        override fun toJson(): String = JSONObject().apply {
            put("type", "pathEstablished")
            put("pathId", pathId)
            put("maxDatagramSize", maxDatagramSize)
        }.toString()
    }

    data class PathLost(val pathId: Long, val reason: String) : PlatformEvent() {
        override fun toJson(): String = JSONObject().apply {
            put("type", "pathLost")
            put("pathId", pathId)
            put("reason", reason)
        }.toString()
    }

    data class NetworkChanged(val kind: String, val available: Boolean) : PlatformEvent() {
        override fun toJson(): String = JSONObject().apply {
            put("type", "networkChanged")
            put("kind", kind)
            put("available", available)
        }.toString()
    }

    data class PermissionChanged(val capability: String, val granted: Boolean) : PlatformEvent() {
        override fun toJson(): String = JSONObject().apply {
            put("type", "permissionChanged")
            put("capability", capability)
            put("granted", granted)
        }.toString()
    }

    data class DeviceStatus(
        val batteryPct: Int?,
        val charging: Boolean,
        val thermallyThrottled: Boolean,
    ) : PlatformEvent() {
        override fun toJson(): String = JSONObject().apply {
            put("type", "deviceStatus")
            put("batteryPct", batteryPct ?: JSONObject.NULL)
            put("charging", charging)
            put("thermallyThrottled", thermallyThrottled)
        }.toString()
    }

    data class LifecycleChanged(val foreground: Boolean) : PlatformEvent() {
        override fun toJson(): String = JSONObject().apply {
            put("type", "lifecycleChanged")
            put("foreground", foreground)
        }.toString()
    }
}

File: apps/mobile/ios/Runner/PlatformEvent.swift
import Foundation

/// Events pushed to the Rust core, normalised to be identical to the Android
/// adapter's output (§10, §90).
///
/// The field names here and in `PlatformEvent.kt` must match exactly — the core
/// has one parser.
enum PlatformEvent {

    case peerAdvertised(kind: String, handle: String, address: String, payload: Data)
    case peerAdvertisementLost(kind: String, handle: String)
    case pathEstablished(pathId: UInt64, maxDatagramSize: Int)
    case pathLost(pathId: UInt64, reason: String)
    case networkChanged(kind: String, available: Bool)
    case audioInterrupted(resumed: Bool)
    case audioRouteChanged(route: String)
    case permissionChanged(capability: String, granted: Bool)
    case deviceStatus(batteryPct: Int?, charging: Bool, thermallyThrottled: Bool)
    case lifecycleChanged(foreground: Bool)

    func jsonString() -> String {
        let object: [String: Any]

        switch self {
        case let .peerAdvertised(kind, handle, address, payload):
            object = [
                "type": "peerAdvertised",
                "kind": kind,
                "handle": handle,
                "address": address,
                "payload": [UInt8](payload).map { Int($0) },
            ]
        case let .peerAdvertisementLost(kind, handle):
            object = ["type": "peerAdvertisementLost", "kind": kind, "handle": handle]
        case let .pathEstablished(pathId, maxDatagramSize):
            object = [
                "type": "pathEstablished",
                "pathId": pathId,
                "maxDatagramSize": maxDatagramSize,
            ]
        case let .pathLost(pathId, reason):
            object = ["type": "pathLost", "pathId": pathId, "reason": reason]
        case let .networkChanged(kind, available):
            object = ["type": "networkChanged", "kind": kind, "available": available]
        case let .audioInterrupted(resumed):
            object = ["type": "audioInterrupted", "resumed": resumed]
        case let .audioRouteChanged(route):
            object = ["type": "audioRouteChanged", "route": route]
        case let .permissionChanged(capability, granted):
            object = [
                "type": "permissionChanged",
                "capability": capability,
                "granted": granted,
            ]
        case let .deviceStatus(batteryPct, charging, thermallyThrottled):
            object = [
                "type": "deviceStatus",
                "batteryPct": batteryPct as Any,
                "charging": charging,
                "thermallyThrottled": thermallyThrottled,
            ]
        case let .lifecycleChanged(foreground):
            object = ["type": "lifecycleChanged", "foreground": foreground]
        }

        guard let data = try? JSONSerialization.data(withJSONObject: object),
              let text = String(data: data, encoding: .utf8)
        else {
            return "{\"type\":\"unknown\"}"
        }
        return text
    }
}
